src/audio/google/google_rtc_audio_processing.c

```diff
@@ -636,118 +636,70 @@ static int google_rtc_audio_processing_prepare(struct processing_module *mod,
-			j = cd->num_aec_reference_channels * cd->aec_reference_frame_index;
-			for (channel = 0; channel < cd->num_aec_reference_channels; ++channel)
-				cd->aec_reference_buffer[j++] = ref[channel];
-
-			ref += audio_stream_get_channels(ref_stream);
-			++cd->aec_reference_frame_index;
-
-			if (cd->aec_reference_frame_index == cd->num_frames) {
-				GoogleRtcAudioProcessingAnalyzeRender_int16(cd->state,
-									    cd->aec_reference_buffer);
-				cd->aec_reference_frame_index = 0;
-			}
-		}
-		num_samples_remaining -= n;
-		ref = audio_stream_wrap(ref_stream, ref);
-	}
-	input_buffers[cd->aec_reference_source].consumed = num_aec_reference_bytes;
-
-	mic_streamb = &input_buffers[cd->raw_microphone_source];
-	mic_stream = mic_streamb->data;
-	out_streamb = &output_buffers[0];
-	out_stream = out_streamb->data;
-
-	src = audio_stream_get_rptr(mic_stream);
-	dst = audio_stream_get_wptr(out_stream);
-
-	frames = input_buffers[cd->raw_microphone_source].size;
-	num_frames_remaining = frames;
-
-	while (num_frames_remaining) {
-		nmax = audio_stream_frames_without_wrap(mic_stream, src);
-		n = MIN(num_frames_remaining, nmax);
-		nmax = audio_stream_frames_without_wrap(out_stream, dst);
-		n = MIN(n, nmax);
-		for (i = 0; i < n; i++) {
-			memcpy_s(&(cd->raw_mic_buffer[cd->raw_mic_buffer_frame_index *
-						      cd->num_capture_channels]),
-				 cd->num_frames * cd->num_capture_channels *
-				 sizeof(cd->raw_mic_buffer[0]), src,
-				 sizeof(int16_t) * cd->num_capture_channels);
-			++cd->raw_mic_buffer_frame_index;
-
-			memcpy_s(dst, cd->num_frames * cd->num_capture_channels *
-				 sizeof(cd->output_buffer[0]),
-				 &(cd->output_buffer[cd->output_buffer_frame_index *
-						     cd->num_capture_channels]),
-				 sizeof(int16_t) * cd->num_capture_channels);
-			++cd->output_buffer_frame_index;
-
-			if (cd->raw_mic_buffer_frame_index == cd->num_frames) {
-				GoogleRtcAudioProcessingProcessCapture_int16(cd->state,
-									     cd->raw_mic_buffer,
-									     cd->output_buffer);
-				cd->output_buffer_frame_index = 0;
-				cd->raw_mic_buffer_frame_index = 0;
-			}
+	int micchan = audio_stream_get_channels(mic);
+	int refchan = audio_stream_get_channels(ref);
+
+	int fmic = audio_stream_get_avail_frames(mic);
+	int fref = audio_stream_get_avail_frames(ref);
+	int frames = MIN(fmic, fref);
+	int n, frames_rem;
 
-			src += audio_stream_get_channels(mic_stream);
-			dst += audio_stream_get_channels(out_stream);
+	for (frames_rem = frames; frames_rem; frames_rem -= n) {
+		n = MIN(frames, cd->num_frames - cd->raw_mic_buffer_frame_index);
+
+		audio_stream_copy_to_linear(mic, 0, cd->raw_mic_buffer,
+					    cd->raw_mic_buffer_frame_index,
+					    n * micchan);
+		audio_stream_copy_to_linear(ref, 0, cd->aec_reference_buffer,
+					    cd->aec_reference_frame_index,
+					    n * refchan);
+		cd->raw_mic_buffer_frame_index += n;
+
+		if (cd->raw_mic_buffer_frame_index >= cd->num_frames) {
```

Comment:
is it possible that: for loop need twice or more time to accumulate data to meet >=?
if possible, then ref and mix offset are always 0 will be wrong when convert to linear.
there should have a index, init as 0, then added index += n;

Reply:
This is the output copy.  It's done all at once at the end of an AEC processing step, so we know the full size and can do it in one step.  Though I guess it's true that in the pathological case the output buffer might be smaller than the AEC processing output size, in which case it would overrun 100% of the time.  Maybe we should detect and warn about that case.

Reply:
ok, if we have confidence on that, we may change to do while loop, and after that check whether still frame_rem still > 0.

Reply:
Generally compilers are good at doing that sort of optimization for themselves (as long as you don't defeated with a function call, etc...), I pasted the generated disassembly of this loop elsewhere in the comments for Guennadi; it looks pretty good to me.